Support Container Attachment

aboch opened the pull request:
This PR contains the API changes to allow agents to request allocation of resources needed to attach a container to a swarm network.

- Adds a new ResourceAllocator service

Signed-off-by: Alessandro Boch <aboch@docker.com>

stevvooe commented on api/dispatcher.proto:
```diff
@@ -48,6 +49,24 @@ service Dispatcher { // maybe dispatch, al likes this
+
+	// ContainerAttachments is a stream of container attachment resources
+	// allocated for the node.
+	rpc ContainerAttachments(ContainerAttachmentsRequest) returns (stream ContainerAttachmentMessage) {
```
Let's call this `Networks` and treat it similar to `Tasks`.

aboch replied:
Wouldn't this be confusing given there is a high level concept of network in swarm, I believe.
Maybe, to conform to the `Tasks` style existing in this interface, would it make sense to just call it `Attachments` ?

aboch replied:
As soon as we finalize on the renaming for `ContainerAttachment`, will check this again.

aboch replied:
I opted for `ExecutorAttachment` in my latest diffs.

stevvooe commented on api/dispatcher.proto:
```diff
@@ -48,6 +49,24 @@ service Dispatcher { // maybe dispatch, al likes this
+
+	// CreateContainerAttachment allows the node to request the resources
+	// allocation needed for a container attachment on the specific node.
+	rpc CreateContainerAttachment(CreateContainerAttachmentRequest) returns (CreateContainerAttachmentResponse) {
```
Should this be through the Control API?

aboch replied:
@mrjana explained it to me this new attachment should not be exposed via the control api, but only to the nodes, given, as I understand, it is going to be for an internal entity node <-> manager (the container attachment which fulfill the requirement gap for running a container to a swarm network), not something which strictly belongs to the regular swarmkit flow for services.

stevvooe replied:
This is not how the division is made. This is still a declaration of "something needs to attach to a network". That needs to be communicated through the control API.

aboch replied:
After this comment, I moved the creation/removal methods to the controlapi, but then the security concern arose, given we would need to open the controlapi to the workers.

So it was discussed offline to go ahead with those methods in the dispatcher API until we find a better place for them.

stevvooe commented on api/specs.proto:
```diff
@@ -241,3 +241,15 @@ message ClusterSpec {
+
+// ContainerAttachmentSpec specifies the allocation information
+// for the container attachment.
+message ContainerAttachmentSpec {
```
Let's see if we can reconcile this type with `NetworkAttachement`.

stevvooe replied:
The key here will be to dispatch `Networks` in the same way we do tasks.

Tasks still attach to networks but they no longer need to carry the entire network object, just a reference.

stevvooe commented on api/specs.proto:
```diff
@@ -241,3 +241,15 @@ message ClusterSpec {
+	string target = 2;
+	
+	// Addresses are the IPv4/IPv6 addresses specified for this attachment.
+	repeated string addresses = 3;
```
Aliases?

aboch replied:
`docker run` and `docker network connect` allow in fact the user to specify the aliases.
But I don't think they have any meaning in the manager territory and we do not need the manger to keep track of them.
This is purely for the resources which needs to be allocated, for now this is only the IP addresses.

stevvooe replied:
How do other containers on the networks know about the aliases?

aboch replied:
No changes it would work as they would today, libnetwork makes sure this info is gossiped among nodes in the network.

stevvooe commented on api/objects.proto:
```diff
@@ -208,3 +208,16 @@ message Cluster {
+// ContainerAttachment specifies the allocation state
+// for the connection of a container to a network.
+message ContainerAttachment {
+	string id = 1 [(gogoproto.customname) = "ID"];
```
What is this identifying?

stevvooe commented on api/specs.proto:
```diff
@@ -241,3 +241,15 @@ message ClusterSpec {
+// for the container attachment.
+message ContainerAttachmentSpec {
+	Annotations annotations = 1 [(gogoproto.nullable) = false];
+	
```
How does this specify which container to attach to?

aboch replied:
The container is not needed. From worker point of view, he needs the manager to allocate the IP resources needed to connect an entity to the network.

aboch commented:
Added the implemntation changes.
Tested these changes in addition to changes in docker/docker and libnetwork:

```
HOST:~$ docker swarm init
Swarm initialized: current node (egszn51lrykxh64u109efxfkh) is now a manager.
HOST1:~$ docker network create -d overlay swarm_nw1
14at7jybsjgosbbul0uxfamkt
HOST1:~$ docker network create -d overlay swarm_nw2
awzkaampy4glcvvssqzx4e4ge
HOST1:~$ docker run -d --name m0 --net swarm_nw1 alpine top
08a19de76aa6f369fcd66b5788a6502716fc94710bb45644a72df6fa9dd43053
HOST1:~$ docker service create --name svc  --network swarm_nw2 alpine top
05c1f6denhozog4hu91jgqnia
HOST1:~$ docker ps
CONTAINER ID        IMAGE               COMMAND             CREATED              STATUS              PORTS               NAMES
9407dfa06b94        alpine:latest       "top"               51 seconds ago       Up 50 seconds                           svc.1.9m6gxnh1csc05tkdimnllcyzi
08a19de76aa6        alpine              "top"               About a minute ago   Up About a minute                       m0
HOST1:~$



HOST2:~$ docker swarm join 192.168.100.126:2377
This node joined a Swarm as a worker.
HOST2:~$ docker run -d --name w0 --net swarm_nw1 alpine top
230d446d545f2c15fcfa2f52eb7fde1cdb6108e3dc29df7b31e5cc36d13ccb2c
HOST2:~$ docker exec w0 ping -c 1 m0
PING m0 (10.0.0.2): 56 data bytes
64 bytes from 10.0.0.2: seq=0 ttl=64 time=0.163 ms

--- m0 ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 0.163/0.163/0.163 ms
HOST2:~$
HOST2:~$ docker network connect swarm_nw2 w0
HOST2:~$ docker exec w0 ping -c 1 svc.1.9m6gxnh1csc05tkdimnllcyzi
PING svc.1.9m6gxnh1csc05tkdimnllcyzi (10.0.1.3): 56 data bytes
64 bytes from 10.0.1.3: seq=0 ttl=64 time=0.742 ms

--- svc.1.9m6gxnh1csc05tkdimnllcyzi ping statistics ---
1 packets transmitted, 1 packets received, 0% packet loss
round-trip min/avg/max = 0.742/0.742/0.742 ms
HOST2:~$ 
```

mavenugo commented:
@aboch thanks for getting this done. 
one of the comments i heard from @aluzzardi & @stevvooe is to create a separate service (not to reuse dispatcher API).  ping @aluzzardi @stevvooe to confirm.

But I guess this PR can be reviewed for the functionality aspects.

stevvooe commented on api/types.proto:
```diff
@@ -359,6 +359,15 @@ message TaskStatus {
 	}
 }
 
+message AttachmentStatus {
```
What is this?

aboch replied:
It is the status of an attachment. This is needed so that actors react on the notification based on the state, which would cause either an allocation or a release of the IP. It may also be needed in the dispatcher code to decide which attachment to send down to the agent, like only the allocated ones.

stevvooe replied:
Reusing task status for this is unwise. If task status represents this resource, then we should use tasks. Although, I think we need a resource allocation model.

aboch replied:
I can add a new enum type, just that I won't be able to call any of its values "ALLOCATED", given that is already used for TaskStatus enum (code generator complains for that). I can get around with this via "RESERVED".  Will make the change.

stevvooe commented on agent/attachment.go:
```diff
@@ -0,0 +1,103 @@
+)
+
+// Listener allow to receive notifications about non executor attachment objects.
+type Listener interface {
```
This package has lots of different types. Could you provide this with a more specific name or leverage the same listener model?

aboch replied:
Yes, Listener is too generic.

stevvooe commented on agent/attachment.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// ExecutorAttachmentManager provides control over executor attachments on this node.
+type ExecutorAttachmentManager interface {
```
Why is this an interface? There is only one implementation and this is not java.

aboch replied:
I was not sure how would docker engine code interact with the agent in order to request the attachment allocation and retrieve the attachments.

My understanding is that for the task mgmt it is all inside the agent and directive comes from agent to docker engine. While in this case the trigger comes from outside the agent.

So I thought of agent providing an attachment management point, through which engine code can do init operations and query info about attachments. 
I feel interfaces work very well for this. So that if in future a new object (volumes?) is allowed to be managed, a new volume management point interface can be provided to the agent's clients.

agent or any other internal structure would be the implementer, of each or of all the management points, but this is not relevant to the interface's user.

stevvooe commented on agent/attachment.go:
```diff
@@ -0,0 +1,103 @@
+
+// CreateAttachment allows the node to request the allocation of
+// resources needed for a container attachment on this node.
+func (a *Agent) CreateAttachment(ctx context.Context, request *api.CreateExecutorAttachmentRequest) (*api.CreateExecutorAttachmentResponse, error) {
```
Don't expose request objects on the exported API.

aboch replied:
ok

stevvooe commented on agent/attachment.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// Register allows clients to register for notifications.
+func (a *Agent) Register(l Listener) (string, error) {
```
I don't understand why agent would implement this interface directly. It has almost nothing to do with the agent.

aboch replied:
This was a practical change.
engine code gets the attachment mgr from agent and operates on that. The fact that agent itself is the implementer of the interface is not known to the user, but I agree it would be better to have a notifier non-exported object that implements the interface.
We can take care of this.

stevvooe commented on api/specs.proto:
```diff
@@ -255,3 +255,18 @@ message ClusterSpec {
+	Annotations annotations = 1 [(gogoproto.nullable) = false];
+
+	// NodeID specifies the node on which the attachment is to be placed
+	string node_id = 2 [(gogoproto.customname) = "NodeID"];
```
Don't we always provide the attachment to the originating node? This should be part of the TLS connection.

aboch replied:
Yes, originally I was getting the nodeid from `ca.RemoteNode(stream.Context())` but moved to this becasue it did not work on the worker, it worked only on the manager, and my goal was to get this work e2e first.
I am happy to derive the nodeID from the connection ctx if it works.

stevvooe commented on agent/attachment.go:
```diff
@@ -0,0 +1,103 @@
+}
+
+// Notifier provides notifications for non Task objects.
+type Notifier interface {
```
How is this different from `Listener`?

aboch replied:
This is interface is not needed, I will remove. Thanks.

stevvooe commented on agent/session.go:
```diff
@@ -27,12 +27,13 @@ var (
+	errs       chan error
+	messages   chan *api.SessionMessage
+	tasks      chan *api.TasksMessage
+	attachmnts chan *api.ExecutorAttachmentMessage
```
`attachments`

stevvooe commented:
I thought we had discussed breaking out network distribution into a `Dispatcher` method `Networks`. This would include a list for networks to create and a list of attachments that aren't bound to a task.

I'd prefer to make these changes correct, rather than trying to fit them in without the correct fixes. The current model is too messy and confusing.

aboch commented:
@stevvooe 
I set my initial goal to first  get the `docker run` and `docker network connect` work end to end on swarm networks. 
With the notion that the further changes to have code comply with swarmkit flow and structures would be needed.
And that these would be additional, incremental changes, which can only be done right through the process of code review from maintainers.

Regarding your request, which I start to understand better now, of dispatcher listing any network attachment (for tasks or for user started containers), implies changes to the Task object I think, where now the agent side code would need to put together the task and the nw attachment before calling the engine/libnetwork code.

If the above is correct, this seems reasonable to me, so that meaningful objects have their indivual disatcher API. 
But based of my current confidence around the swam code, this seems at first a big refactor change,  and  I am not that confident to have it part his PR.

stevvooe commented:
> But based of my current confidence around the swam code, this seems at first a big refactor change, and I am not that confident to have it part his PR.

My suggestions shouldn't result in a very large refactor. We just need to move a few fields around. The big change will be separating out the creation of networks from the main task flow. This will immensely simplify the object model of networks and tasks.

I don't think we can merge this functionality _without_ changes here. Without doing so, we will leak networks under failure modes. Before this change, a networks life cycle was bound by the life cycle of the set of referencing tasks. With this change, we no longer have that lifecycle hook. Networks must have their own lifecycle.

aboch commented:
@stevvooe 
I dropped the ExecutorAttachment object and specs and reused the NetworkAttachment object and added a spec for it.
The dispatcher API now will give a stream of NetworkAttachment objects.

For now the stream will only include the attachments that were created from engine, to connect containers to swarm networks. But at least this should be inline with the future changes you had in mind where also the tasks' attachments will be dispatched via this stream.

Please take a look when you get a chance.
Thanks.

stevvooe commented:
> But at least this should be inline with the future changes you had in mind where also the tasks' attachments will be dispatched via this stream.

We need to make these fixes now. The current changes aren't the right model for GA and its clear the networking model is insufficient for our needs.

stevvooe commented on api/specs.proto:
```diff
@@ -255,3 +255,21 @@ message ClusterSpec {
+
+// NetworkAttachmentSpec specifies the allocation information
+// for the network attachment.
+message NetworkAttachmentSpec {
```
This should actually be `NetworkAttachmentConfig`. `Spec` should only _ever_ be user input.

aboch commented:
>  Before this change, a networks life cycle was bound by the life cycle of the set of referencing tasks. With this change, we no longer have that lifecycle hook. Networks must have their own lifecycle.

Networks today have their own lifecyle independent of Tasks.
I think you are referring to NetworkAttachment objects which lifecyle is in fact are bound to the Task.

With this change we are allowing engine to create its own NetworkAttachment objects for its own containers user wants to connect to the swarm networks. Yes, their lifecyle is bound to the user started containers. So now we would have NetworkAttachment objects managed by swarm (linked from Task lifecycle) and NetworkAttachment objects controlled by engine (bound to user container lifecyle). 

But the only reason why we need swarm mgr to know about these attachments is just so that it can allocate the IP adresses, so that it won't give to tasks IPs which are being used by user started containers. We do not really need to pass this as NetworkAttachment, which is why initially I had a separate ContainerAttachment object.

aluzzardi commented on manager/dispatcher/attachment.go:
```diff
@@ -0,0 +1,75 @@
+// - Returns `InvalidArgument` if attachment ID is not provided.
+// - Returns `NotFound` if the attachment is not found.
+// - Returns an error if the deletion fails.
+func (d *Dispatcher) RemoveNetworkAttachment(ctx context.Context, request *api.RemoveNetworkAttachmentRequest) (*api.RemoveNetworkAttachmentResponse, error) {
```
This allows any worker to delete any attachment, including other workers'

aboch replied:
Thanks. Will make sure a worker can only delete attachments created on the same node

aluzzardi commented on manager/dispatcher/attachment.go:
```diff
@@ -0,0 +1,75 @@
+// - Returns `InvalidArgument` if the Spec is malformed.
+// - Returns `NotFound` if the Network is not found.
+// - Returns an error if the creation fails.
+func (d *Dispatcher) CreateNetworkAttachment(ctx context.Context, request *api.CreateNetworkAttachmentRequest) (*api.CreateNetworkAttachmentResponse, error) {
```
There's some rate limiting code used in the dispatcher. This function should also make use of it

stevvooe replied:
That is session-based rate limiting. I am not sure it generically applies.

What issue with rate limiting do you see?

stevvooe commented on agent/attachment.go:
```diff
@@ -0,0 +1,97 @@
+type NetworkAttachmentManager interface {
+	// CreateAttachment allows the node to request the allocation of resources
+	// needed for a network attachment on this node.
+	CreateAttachment(context.Context, *api.NetworkAttachmentConfig) (string, error)
```
Move this to worker:

```
AttachNetwork
LeaveNetwork
```

We'll work on registration in a bit.

aluzzardi commented:
@aboch Please rebase

mavenugo commented:
@aluzzardi @icecrime  thanks to @aboch we now have a full e2e integrated PR : https://github.com/docker/docker/pull/25263

dperny commented:
Is this PR superseded by #1373 ? @mrjana @aboch ?

mrjana commented:
@dperny Yes.
